File: app/src/main/java/com/major/qr/viewmodels/AttendanceViewModel.java
package com.major.qr.viewmodels;

import static com.major.qr.ui.LoginActivity.URL;
import static com.major.qr.ui.LoginActivity.requestQueue;

import android.util.Log;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.android.volley.Request;
import com.android.volley.toolbox.JsonArrayRequest;
import com.android.volley.toolbox.JsonObjectRequest;
import com.major.qr.models.Attendance;
import com.major.qr.ui.LoginActivity;
import com.major.qr.volley.VolleySingleton;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class AttendanceViewModel extends ViewModel {
    public final static String ATTENDANCE_API_URL = URL + "/attendance/";
    public final String TAG = AttendanceViewModel.class.getSimpleName();
    MutableLiveData<ArrayList<Attendance>> list;

    public MutableLiveData<ArrayList<Attendance>> getAttendances() {
        if (list == null) list = new MutableLiveData<>();
        loadAttendanceList();
        return list;
    }

    private void loadAttendanceList() {
        final String url = URL + "/attendance/get";
        JsonArrayRequest request = new JsonArrayRequest(Request.Method.GET, url, null, jsonArray -> {
            Log.d(TAG, "response: " + jsonArray.toString());
            ArrayList<Attendance> attendanceArrayList = new ArrayList<>();
            try {
                for (int i = 0; i < jsonArray.length(); ++i) {
                    JSONObject object = jsonArray.getJSONObject(i);
                    Log.d(TAG, "loadAttendanceList: " + object);
                    if (!object.has("name")) continue;

                    attendanceArrayList.add(new Attendance(object.getString("id"),
                            object.getString("name"),
                            object.getString("totalAttenders"),
                            object.getString("creationDate")));
                }
                list.postValue(attendanceArrayList);
            } catch (JSONException e) {
                throw new RuntimeException(e);
            }
        }, error -> Log.e(TAG, "error: " + error.toString())) {
            @Override
            public Map<String, String> getHeaders() {
                return new HashMap<String, String>() {{
                    put("Authorization", LoginActivity.ACCESS_TOKEN);
                }};
            }
        };
        requestQueue.add(request);
    }

    public MutableLiveData<String> markAttendance(String uid, String attendanceId) {
        MutableLiveData<String> mutableLiveData = new MutableLiveData<>();
        JSONObject params = new JSONObject() {{
            try {
                put("attendanceId", attendanceId);
                put("attendersId", LoginActivity.USERID);
                put("displayName", LoginActivity.NAME);
                put("email", LoginActivity.EMAIL);
            } catch (JSONException e) {
                throw new RuntimeException(e);
            }
        }};
        final String url = URL + "/attendance/mark?uid=" + uid;
        Log.d(TAG, "markAttendance: " + url);
        JsonObjectRequest jsonObjectRequest = new JsonObjectRequest(Request.Method.POST, url, params,
                response -> {
                    mutableLiveData.postValue(String.valueOf(response));
                    Log.d(TAG, response.toString());
                },
                error -> {
                    mutableLiveData.postValue(null);
                    Log.d(TAG, "uploadDoc: No internet Connection");
                }) {
            @Override
            public Map<String, String> getHeaders() {
                return new HashMap<String, String>() {{
                    put("accept", "*/*");
                    put("Authorization", LoginActivity.ACCESS_TOKEN);
                    put("Content-Type", "application/json");
                }};
            }
        };
        requestQueue.add(jsonObjectRequest.setRetryPolicy(VolleySingleton.getRetryPolicy()));
        return mutableLiveData;
    }

    public MutableLiveData<JSONObject> createAttendance(String name) {
        MutableLiveData<JSONObject> res = new MutableLiveData<>();
        final String url = ATTENDANCE_API_URL + "create?name=" + name;
        Log.d(TAG, "createAttendance: " + url);
        JsonObjectRequest request = new JsonObjectRequest(Request.Method.GET, url, null,
                response -> {
                    loadAttendanceList();
                    res.postValue(response);
                }, error -> Log.d(TAG, "uploadDoc: No internet Connection")) {
            @Override
            public Map<String, String> getHeaders() {
                return new HashMap<String, String>() {{
                    put("accept", "*/*");
                    put("Authorization", LoginActivity.ACCESS_TOKEN);
                }};
            }
        };
        requestQueue.add(request.setRetryPolicy(VolleySingleton.getRetryPolicy()));
        return res;
    }

    public MutableLiveData<JSONObject> deleteAttendance(String attendanceId) {
        MutableLiveData<JSONObject> liveData = new MutableLiveData<>();
        final String url = URL + "/attendance/delete?attendanceId=" + attendanceId;
        JsonObjectRequest request = new JsonObjectRequest(Request.Method.DELETE, url, null,
                response -> {
                    loadAttendanceList();
                    liveData.postValue(response);
                }, error -> {
            liveData.postValue(null);
            Log.e(TAG, "error: " + error.toString());
        }) {
            @Override
            public Map<String, String> getHeaders() {
                return new HashMap<String, String>() {{
                    put("Authorization", LoginActivity.ACCESS_TOKEN);
                }};
            }
        };
        requestQueue.add(request);
        return liveData;
    }
}

File: app/src/main/java/com/major/qr/viewmodels/AttendanceViewModel.kt
package com.major.qr.viewmodels

import android.util.Log
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.android.volley.Response
import com.android.volley.VolleyError
import com.android.volley.toolbox.JsonArrayRequest
import com.android.volley.toolbox.JsonObjectRequest
import com.major.qr.models.Attendance
import com.major.qr.ui.LoginActivity
import com.major.qr.volley.VolleySingleton
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

class AttendanceViewModel : ViewModel() {
    val TAG = AttendanceViewModel::class.java.simpleName
    var list: MutableLiveData<ArrayList<Attendance>>? = null
    val attendances: MutableLiveData<ArrayList<Attendance>>
        get() {
            if (list == null) list = MutableLiveData()
            loadAttendanceList()
            return list as MutableLiveData<ArrayList<Attendance>>
        }

    private fun loadAttendanceList() {
        val url = LoginActivity.URL + "/attendance/get"
        val request: JsonArrayRequest = object :
            JsonArrayRequest(Method.GET, url, null, Response.Listener { jsonArray: JSONArray ->
                Log.d(TAG, "response: $jsonArray")
                val attendanceArrayList = ArrayList<Attendance>()
                try {
                    for (i in 0 until jsonArray.length()) {
                        val `object` = jsonArray.getJSONObject(i)
                        Log.d(TAG, "loadAttendanceList: $`object`")
                        if (!`object`.has("name")) continue
                        attendanceArrayList.add(
                            Attendance(
                                `object`.getString("id"),
                                `object`.getString("name"),
                                `object`.getString("totalAttenders"),
                                `object`.getString("creationDate")
                            )
                        )
                    }
                    list!!.postValue(attendanceArrayList)
                } catch (e: JSONException) {
                    throw RuntimeException(e)
                }
            }, Response.ErrorListener { error: VolleyError -> Log.e(TAG, "error: $error") }) {
            override fun getHeaders(): HashMap<String?, String?> {
                return object : HashMap<String?, String?>() {
                    init {
                        put("Authorization", LoginActivity.ACCESS_TOKEN)
                    }
                }
            }
        }
        LoginActivity.requestQueue?.add(request)
    }

    fun markAttendance(uid: String, attendanceId: String?): MutableLiveData<String?> {
        val mutableLiveData = MutableLiveData<String?>()
        val params: JSONObject = object : JSONObject() {
            init {
                try {
                    put("attendanceId", attendanceId)
                    put("attendersId", LoginActivity.USERID)
                    put("displayName", LoginActivity.NAME)
                    put("email", LoginActivity.EMAIL)
                } catch (e: JSONException) {
                    throw RuntimeException(e)
                }
            }
        }
        val url = LoginActivity.URL + "/attendance/mark?uid=" + uid
        Log.d(TAG, "markAttendance: $url")
        val jsonObjectRequest: JsonObjectRequest = object : JsonObjectRequest(
            Method.POST, url, params,
            Response.Listener<JSONObject> { response: JSONObject ->
                mutableLiveData.postValue(response.toString())
                Log.d(TAG, response.toString())
            },
            Response.ErrorListener { error: VolleyError? ->
                mutableLiveData.postValue(null)
                Log.d(TAG, "uploadDoc: No internet Connection")
            }) {
            override fun getHeaders(): HashMap<String?, String?> {
                return object : HashMap<String?, String?>() {
                    init {
                        put("accept", "*/*")
                        put("Authorization", LoginActivity.ACCESS_TOKEN)
                        put("Content-Type", "application/json")
                    }
                }
            }
        }
        LoginActivity.requestQueue?.add(jsonObjectRequest.setRetryPolicy(VolleySingleton.getRetryPolicy()))
        return mutableLiveData
    }

    fun createAttendance(name: String): MutableLiveData<JSONObject> {
        val res = MutableLiveData<JSONObject>()
        val url = ATTENDANCE_API_URL + "create?name=" + name
        Log.d(TAG, "createAttendance: $url")
        val request: JsonObjectRequest = object : JsonObjectRequest(Method.GET,
            url,
            null,
            Response.Listener { response: JSONObject ->
                loadAttendanceList()
                res.postValue(response)
            },
            Response.ErrorListener { error: VolleyError? ->
                Log.d(
                    TAG,
                    "uploadDoc: No internet Connection"
                )
            }) {
            override fun getHeaders(): HashMap<String?, String?> {
                return object : HashMap<String?, String?>() {
                    init {
                        put("accept", "*/*")
                        put("Authorization", LoginActivity.ACCESS_TOKEN)
                    }
                }
            }
        }
        LoginActivity.requestQueue?.add(request.setRetryPolicy(VolleySingleton.getRetryPolicy()))
        return res
    }

    fun deleteAttendance(attendanceId: String): MutableLiveData<JSONObject?> {
        val liveData = MutableLiveData<JSONObject?>()
        val url = LoginActivity.URL + "/attendance/delete?attendanceId=" + attendanceId
        val request: JsonObjectRequest = object : JsonObjectRequest(Method.DELETE, url, null,
            Response.Listener { response: JSONObject? ->
                loadAttendanceList()
                liveData.postValue(response)
            }, Response.ErrorListener { error: VolleyError ->
                liveData.postValue(null)
                Log.e(TAG, "error: $error")
            }) {
            override fun getHeaders(): HashMap<String?, String?> {
                return object : HashMap<String?, String?>() {
                    init {
                        put("Authorization", LoginActivity.ACCESS_TOKEN)
                    }
                }
            }
        }
        LoginActivity.requestQueue?.add(request)
        return liveData
    }

    companion object {
        const val ATTENDANCE_API_URL = LoginActivity.URL + "/attendance/"
    }
}